tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml

```diff
@@ -75,40 +78,250 @@ stages:
+      continueOnError: true
+
+    - script: |
+        JEST_JUNIT_OUTPUT_FILE=$(Build.SourcesDirectory)/js/react_native/e2e/ios-test-results.xml \
+          detox test --record-logs all \
+                     --configuration ios.sim.release \
+                     --loglevel trace \
+                     --take-screenshots failing
+      workingDirectory: '$(Build.SourcesDirectory)/js/react_native/e2e'
+      displayName: Run React Native Detox iOS e2e Tests
+
+    - script: |
+        set -e
+        DEVICE_ID_FILE="$(Build.BinariesDirectory)/ios_simulator_device_id.txt"
+        if [[ -f "${DEVICE_ID_FILE}" ]]; then
+          xcrun simctl delete "$(cat "${DEVICE_ID_FILE}")"
+          rm "${DEVICE_ID_FILE}"
+        fi
+      displayName: Stop iOS Simulator
+      condition: always()
+
+    - task: PublishTestResults@2
+      inputs:
+        testResultsFiles: '$(Build.SourcesDirectory)/js/react_native/e2e/ios-test-results.xml'
+        failTaskOnFailedTests: true
+        testRunTitle: 'React Native Detox iOS e2e Test Results'
+      condition: succeededOrFailed()
+      displayName: Publish React Native Detox iOS e2e Test Results
+
+
+
+    - script: |
+        git restore .
+      workingDirectory: '$(Build.SourcesDirectory)/js'
+      displayName: Restore git changes
+
+    - task: PowerShell@2
+      inputs:
+        filePath: '$(Build.SourcesDirectory)/tools/ci_build/github/js/pack-npm-packages.ps1'
+        arguments: '"${{parameters.NpmPackagingMode}}" $(Build.SourcesDirectory) react_native'
+        workingDirectory: '$(Build.SourcesDirectory)'
+        errorActionPreference: stop
+      displayName: Pack NPM packages
+
+    - task: CopyFiles@2
+      inputs:
+        sourceFolder: $(Build.SourcesDirectory)/js/common
+        contents: onnxruntime-common*.tgz
+        targetFolder: $(Build.ArtifactStagingDirectory)
+      displayName: 'Create Artifacts onnxruntime-common'
+
+    - task: CopyFiles@2
+      inputs:
+        sourceFolder: $(Build.SourcesDirectory)/js/react_native
+        contents: onnxruntime-react-native*.tgz
+        targetFolder: $(Build.ArtifactStagingDirectory)
+      displayName: Create Artifacts onnxruntime-react-native
+
+    - task: PublishPipelineArtifact@1
+      inputs:
+        artifact: ios_e2e_test_logs_$(Build.BuildId)_$(Build.BuildNumber)_$(System.JobAttempt)
+        targetPath: '$(Build.SourcesDirectory)/js/react_native/e2e/artifacts'
+      condition: succeededOrFailed()
+      displayName: Publish React Native Detox E2E test logs
+
+    - task: PublishPipelineArtifact@0
+      inputs:
+        artifactName: 'NPM_packages_ios'
+        targetPath: '$(Build.ArtifactStagingDirectory)'
+      displayName: Publish Pipeline Artifact
+
+    - template: explicitly-defined-final-tasks.yml
+
+
+- stage: ReactNative_CI_Android
+  displayName: ReactNative_CI_Android
   dependsOn:
-  - Build_Packages
+  - Build_Android_Packages
   jobs:
-  - job: ReactNative_CI
+  - job: ReactNative_CI_Android
```

Comment:
Could you use Linux machines to build Android?

Reply:
They are more stable, and we have more capacities there.

Reply:
If you do it in a new PR if you'd like.

Reply:
@edgchen1 That is a question for Edward.

Reply:
worth trying it. IIRC, we used to only be able to run the android emulator on Mac agents.

Reply:
Now it runs well on Linux.